File: frontend/src/api/request-logs.ts
import { requestJson } from "./client";
import type { ApiType } from "./providers";

export interface RequestLogSummary {
  request_id: string;
  gateway_key_id: string | null;
  api_type: ApiType;
  model: string | null;
  alias: string | null;
  provider: string | null;
  endpoint: string | null;
  status_code: number | null;
  latency_ms: number | null;
  client_ip: string | null;
  user_agent: string | null;
  request_content_type: string | null;
  response_content_type: string | null;
  prompt_tokens: number | null;
  completion_tokens: number | null;
  total_tokens: number | null;
  created_at: string;
}

export interface RequestLog {
  request_id: string;
  gateway_key_id: string | null;
  api_type: ApiType;
  model: string | null;
  alias: string | null;
  provider: string | null;
  endpoint: string | null;
  status_code: number | null;
  latency_ms: number | null;
  client_ip: string | null;
  user_agent: string | null;
  request_body: number[] | null; // Vec<u8> is array of numbers in JSON
  response_body: number[] | null;
  request_content_type: string | null;
  response_content_type: string | null;
  prompt_tokens: number | null;
  completion_tokens: number | null;
  total_tokens: number | null;
  created_at: string;
}

export interface RequestLogFilter {
  model?: string;
  alias?: string;
  provider?: string;
  client_ip?: string;
}

export const fetchRequestLogs = async (
  limit: number = 20,
  offset: number = 0,
  filter: RequestLogFilter = {},
): Promise<{ data: RequestLogSummary[]; total: number }> => {
  const params = new URLSearchParams({
    limit: limit.toString(),
    offset: offset.toString(),
  });

  if (filter.model) params.append("model", filter.model);
  if (filter.alias) params.append("alias", filter.alias);
  if (filter.provider) params.append("provider", filter.provider);
  if (filter.client_ip) params.append("client_ip", filter.client_ip);

  return requestJson<{ data: RequestLogSummary[]; total: number }>(
    `/request-logs?${params.toString()}`,
  );
};

export const fetchRequestLog = async (id: string): Promise<RequestLog> => {
  return requestJson<RequestLog>(`/request-logs/${id}`);
};

File: backend/src/handlers/request_logs.rs
use axum::{
    Json,
    extract::{Path, Query, State},
};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use crate::{
    db::request_logs::{RequestLog, RequestLogFilter, RequestLogSummary},
    error::{AppError, AppResult},
    services::logging::{self},
    state::AppState,
};

#[derive(Deserialize)]
pub struct ListRequestLogsQuery {
    pub limit: Option<i64>,
    pub offset: Option<i64>,
    pub request_id: Option<Uuid>,
    pub model: Option<String>,
    pub alias: Option<String>,
    pub provider: Option<String>,
    pub status_code: Option<i32>,
    pub client_ip: Option<String>,
}

#[derive(Serialize)]
pub struct ListRequestLogsResponse {
    pub data: Vec<RequestLogSummary>,
    pub total: i64,
}

pub async fn list_request_logs(
    State(state): State<AppState>,
    Query(query): Query<ListRequestLogsQuery>,
) -> AppResult<Json<ListRequestLogsResponse>> {
    let filter = RequestLogFilter {
        limit: query.limit.or(Some(20)),
        offset: query.offset.or(Some(0)),
        request_id: query.request_id,
        model: query.model,
        alias: query.alias,
        provider: query.provider,
        status_code: query.status_code,
        client_ip: query.client_ip,
    };

    let logs = logging::fetch_request_logs(&state.pool, &filter).await?;
    let total = logging::count_request_logs(&state.pool, &filter).await?;

    Ok(Json(ListRequestLogsResponse { data: logs, total }))
}

pub async fn get_request_log(
    State(state): State<AppState>,
    Path(request_id): Path<Uuid>,
) -> AppResult<Json<RequestLog>> {
    let log = logging::fetch_request_log_detail(&state.pool, request_id).await?;

    match log {
        Some(log) => Ok(Json(log)),
        None => Err(AppError::NotFound),
    }
}
